datafusion/scheduler/src/task.rs

```diff
@@ -0,0 +1,225 @@
+                waker: Arc::new(TaskWaker {
+                    query: Arc::downgrade(&query),
+                    wake_count: AtomicUsize::new(1),
+                    pipeline: pipeline_idx,
+                    partition,
+                }),
+            });
+        }
+    }
+}
+
+/// A [`Task`] identifies an output partition within a given pipeline that may be able to
+/// make progress. The [`Scheduler`][super::Scheduler] maintains a list of outstanding
+/// [`Task`] and distributes them amongst its worker threads.
+///
+/// A [`Query`] is considered completed when it has no outstanding [`Task`]
+pub struct Task {
+    /// Maintain a link to the [`Query`] this is necessary to be able to
+    /// route the output of the partition to its destination, and also because
+    /// when [`Query`] is dropped it signals completion of query execution
+    query: Arc<Query>,
+
+    /// A [`ArcWake`] that can be used to re-schedule this [`Task`] for execution
+    waker: Arc<TaskWaker>,
+}
+
+impl std::fmt::Debug for Task {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let output = self.query.pipelines()[self.waker.pipeline].output;
+
+        f.debug_struct("Task")
+            .field("pipeline", &self.waker.pipeline)
+            .field("partition", &self.waker.partition)
+            .field("output", &output)
+            .finish()
+    }
+}
+
+impl Task {
+    /// Call [`Pipeline::poll_partition`] attempting to make progress on query execution
+    pub fn do_work(self) {
+        assert!(is_worker(), "Task::do_work called outside of worker pool");
+        if self.query.is_cancelled() {
+            return;
+        }
+
+        // Capture the wake count prior to calling [`Pipeline::poll_partition`]
+        // this allows us to detect concurrent wake ups and handle them correctly
+        //
+        // We aren't using the wake count to synchronise other memory, and so can
+        // use relaxed memory ordering
+        let wake_count = self.waker.wake_count.load(Ordering::Relaxed);
+
+        let node = self.waker.pipeline;
+        let partition = self.waker.partition;
+
+        let waker = futures::task::waker_ref(&self.waker);
+        let mut cx = Context::from_waker(&*waker);
+
+        let pipelines = self.query.pipelines();
+        let routable = &pipelines[node];
+        match routable.pipeline.poll_partition(&mut cx, partition) {
+            Poll::Ready(Some(Ok(batch))) => {
+                trace!("Poll {:?}: Ok: {}", self, batch.num_rows());
+                match routable.output {
+                    Some(link) => {
+                        trace!(
+                            "Publishing batch to pipeline {:?} partition {}",
+                            link,
+                            partition
+                        );
+                        pipelines[link.pipeline]
+                            .pipeline
+                            .push(batch, link.child, partition)
```

Comment:
Is it correct that this call may do a non trivial amount of work (aka this is the push / keep working on the batch in the same thread that produced it)?

Reply:
Yes, which does create an interesting situation as the next output can't be polled until the last batch has been completely processed 🤔

I think this will all shake out once we are consistently push-based

Reply:
> Yes, which does create an interesting situation as the next output can't be polled until the last batch has been completely processed

But this should be ok because all the other threads should be actively doing something if there is work to do, right?

Reply:
Yes, there should be plenty of parallelism to go around in most cases 😃